### PR TITLE
Fixed the tenant switching after timeout

### PR DESCRIPTION
**Pull request by RyanL1997**
Signed-off-by: Ryan Liang <jiallian@amazon.com>

### Description
Fix the issue of user tenant automatically switch back to Private/Global by default after session timeout.

### Category
Bug fix

### Issues Resolved
* Resolved https://github.com/opensearch-project/security-dashboards-plugin/issues/1084

### Check List
- [x] New functionality includes testing
- [ ] ~New functionality has been documented~
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by cliu123 — DISMISSED** [review]

**Review by cwperks — APPROVED** [review]
Looks good. Thanks @RyanL1997!

**Comment by anijain-Amazon** [comment]
@cwperks Please add the backport label for 1.x as well.

File: public/plugin.ts [hunk 1]
```diff
@@ -150,19 +151,21 @@ export class SecurityPlugin implements Plugin<SecurityPluginSetup, SecurityPlugi
       // logout the user when getting 401 unauthorized, e.g. when session timed out.
       core.http.intercept({
         responseError: (httpErrorResponse, controller) => {
```
Comment (peternied) [inline public/plugin.ts:153]: Please add a test to verify the new behavior

**Review by peternied — CHANGES_REQUESTED** [review]
Need to add tests to verify this change

File: public/plugin.ts (hunk 1 above)
Comment (RyanL1997) [reply public/plugin.ts:153]: Hi @peternied, thanks for the reply! I have tested it locally by viewing the session storage value after the `ttl` session got expired. Im working on adding some of the test to this PR now.

**Comment by cliu123** [comment]
@RyanL1997 Could you please add the manual testing in the PR description for now and add ITs in a follow up PR if adding ITs will take time?

File: public/utils/logout-utils.tsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        window.location.pathname.toLowerCase().includes(LOGIN_PAGE_URI) ||
+        window.location.pathname.toLowerCase().includes(CUSTOM_ERROR_PAGE_URI)
+        )
+        ) {
+        if (logoutUrl) {
+            thisWindow.location.href = logoutUrl;
+        } else {
+            // when session timed out, user credentials in cookie are wiped out
+                // refres the page will direct the user to go through login process
+            window.location.reload();
```
Comment (peternied) [inline public/utils/logout-utils.tsx:38]: Please use `thisWindow` for this call too

File: public/apps/account/test/plugin.test.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+}));
+
+describe('Session timeout', () => {
+  const fakeError = {
+    response: {
+      status: 401,
+    },
+  };
+
+  it('Should call setShouldShowTenantPopup on session timeout', () => {
```
Comment (peternied) [inline public/apps/account/test/plugin.test.tsx:31]: Please also verify the `thisWindow.location.href` or `thisWindow.location.reload()` calls (you can use another mock for this.

File: public/apps/account/test/plugin.test.tsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    const sessionTimeoutFn = interceptError(LOGIN_PAGE_URI, window);
+    sessionTimeoutFn(fakeError, null);
+    expect(setShouldShowTenantPopup).toBeCalledTimes(1);
```
Comment (peternied) [inline public/apps/account/test/plugin.test.tsx:34]: It would also be good to test a negative test cases and be sure the popup isn't called on the non 401 case

**Review by peternied — CHANGES_REQUESTED** [review]
Thanks for adding the tests - a couple of comments about theses and I think we should add another test case

File: public/utils/logout-utils.tsx [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  HttpInterceptorResponseError,
+  IHttpInterceptController,
+} from '../../../../src/core/public';
+import { CUSTOM_ERROR_PAGE_URI, LOGIN_PAGE_URI } from '../../common';
+
+export function interceptError(logoutUrl: string, thisWindow: Window): any {
+    return (httpErrorResponse: HttpInterceptorResponseError, _: IHttpInterceptController) => {
+    if (httpErrorResponse.response?.status === 401) {
+        setShouldShowTenantPopup(null);
+    if (
```
[the remark this answers is not in the document]
Comment (RyanL1997) [reply public/utils/logout-utils.tsx]: Got it! I'm fixing the format according to the lint in unit test.

File: public/utils/logout-utils.tsx (hunk 2 above)
Comment (RyanL1997) [reply public/utils/logout-utils.tsx]: I just fixed this. Thanks for the review!

File: public/utils/logout-utils.tsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+} from '../../../../src/core/public';
+import { CUSTOM_ERROR_PAGE_URI, LOGIN_PAGE_URI } from '../../common';
+
+export function interceptError(logoutUrl: string, thisWindow: Window): any {
+  return (httpErrorResponse: HttpInterceptorResponseError, _: IHttpInterceptController) => {
+    if (httpErrorResponse.response?.status === 401) {
+      setShouldShowTenantPopup(null);
+      if (
+        !(
+          window.location.pathname.toLowerCase().includes(LOGIN_PAGE_URI) ||
```
Comment (cwperks) [inline public/utils/logout-utils.tsx:29]: Should this line and the one below also be `thisWindow`?

Comment (RyanL1997) [reply public/utils/logout-utils.tsx]: Hi @cwperks, thx for the review! You are right. I just fixed this.

File: public/apps/account/test/plugin.test.tsx (hunk 4 above)
Comment (RyanL1997) [reply public/apps/account/test/plugin.test.tsx:40]: Thx for the suggestion @peternied! I have updated it with a negative test cases for this one.

File: public/apps/account/test/plugin.test.tsx (hunk 3 above)
Comment (RyanL1997) [reply public/apps/account/test/plugin.test.tsx]: @peternied thanks for this review! This is out of the scope of the PR that fixes the switching tenant after timeout as the fix has been tested. We can create a follow-up PR for this if needed.

File: public/utils/logout-utils.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (RyanL1997) [reply public/utils/logout-utils.tsx]: I just fixed this one. Thank you for the review.

**Review by peternied — APPROVED** [review]
Looks good, I think you'll need to rebase 2.4.0 has merged otherwise intergration tests aren't going to work

**Comment by peternied** [comment]
See https://github.com/opensearch-project/security-dashboards-plugin/pull/1096

**Comment by RyanL1997** [comment]
> @RyanL1997 Could you please add the manual testing in the PR description for now and add ITs in a follow up PR if adding ITs will take time?

@cliu123 Sorry for the late reply. I have been working on the unit tests and I just updated the PR with automation tests cases.

**Comment by RyanL1997** [comment]
> Looks good, I think you'll need to rebase 2.4.0 has merged otherwise intergration tests aren't going to work

@peternied Got it! Thank you again for the review!

**Review by DarshitChanpura — CHANGES_REQUESTED** [review]
Looks good Ryan. Can you please rebase it with https://github.com/cwperks/security-dashboards-plugin/tree/increment-to-2.4.0.0 so ITs can pass

**Comment by RyanL1997** [comment]
> Looks good Ryan. Can you please rebase it with https://github.com/cwperks/security-dashboards-plugin/tree/increment-to-2.4.0.0 so ITs can work

@DarshitChanpura I just did the rebase and pushed. Thx DC!

**Comment by cliu123** [comment]
@RyanL1997 The rebase was not done properly. The version bump changes are not supposed to be present in this PR.

**Comment by RyanL1997** [comment]
> @RyanL1997 The rebase was not done properly. The version bump changes are not supposed to be present in this PR.

Hi @cliu123, ty for the review, and I just fixed this.

**Review by DarshitChanpura — APPROVED** [review]
Ty @RyanL1997 !

**Review by peternied — APPROVED** [review]
